### PR TITLE
feat: add a pgproxy for rewriting dsn strings

### PR DESCRIPTION
**Pull request by jvmakine**
Still missing implementation for cancellation, which I will add as a followup.

Will also add custom DSN rewriters, like IAM based authenticator to the `ftl-proxy-pg`

### CODE REVIEW

File: internal/pgproxy/pgproxy.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+type PgProxy struct {
+	// listenAddress is the address to listen on for incoming connections.
+	listenAddress string
+	// connectionStringFn is a function that constructs a new connection string from parameters of the incoming connection.
+	connectionStringFn func(ctx context.Context, params map[string]string) (string, error)
+}
+
+// NewPgProxy creates a new PgProxy.
+func NewPgProxy(address string, connectionFn func(ctx context.Context, params map[string]string) (string, error)) *PgProxy {
```
Comment (alecthomas) [inline internal/pgproxy/pgproxy.go:22]: nit: typically for a package with a single type, I'd just use `New()` as the extra name is redundant and stutters: `pgproxy.NewPgProxy`

Comment (alecthomas) [reply internal/pgproxy/pgproxy.go:22]: Can you document what "params" is?

Comment (alecthomas) [reply internal/pgproxy/pgproxy.go:22]: Or maybe just move the docs from the type fields to the function, as that's public.

File: internal/pgproxy/pgproxy.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+		connectionStringFn: connectionFn,
+	}
+}
+
+// Start starts the proxy.
+func (p *PgProxy) Start(ctx context.Context) error {
+	logger := log.FromContext(ctx)
+
+	if p.connectionStringFn == nil {
+		return fmt.Errorf("ConnectionStringFn is required")
```
Comment (alecthomas) [inline internal/pgproxy/pgproxy.go:34]: I'd probably just omit these checks, but if you do want to keep them just panic - this is a misuse of the API and programmer error

**Review by alecthomas — APPROVED** [review]
Very cool.

**Comment by alecthomas** [comment]
Needs some tests though.

File: internal/pgproxy/pgproxy.go (hunk 2 above)
Comment (jvmakine) [reply internal/pgproxy/pgproxy.go:34]: Ah, this was leftover before I added `New` Will remove, as it is not needed anymore

**Pull request merged**